Document/0x05f-Testing-Local-Authentication.md

```diff
@@ -82,21 +84,66 @@ Make sure that the unlocked key is used during the application flow. For example
 
 An even more secure option is using asymmetric cryptography. Here, the mobile app creates an asymmetric key pair in the KeyStore and enrolls the public key on the server backend. Later transactions are then signed with the private key and verified by the server using the public key. The advantage of this is that transactions can be signed using KeyStore APIs without ever extracting the private key from the KeyStore. Consequently, it is impossible for attackers to obtain the key from memory dumps or by using instrumentation.
 
+#### Static Analysis
+
 Note that there are quite some SDKs provided by vendors, which should provide biometric support, but which have their own insecurities. Be very cautious when using third party SDKs to handle sensitive authentication logic.
 
-#### Static Analysis
+The following sections explain the different biometric authentication classes.
+
+##### Biometric Library
+
+A library called [Biometric](https://developer.android.com/jetpack/androidx/releases/biometric "Biometric library for Android") is available for Android that offers a compatibility version of the `BiometricPrompt` and `BiometricManager` APIs, as implemented in Android 10, with full feature support back to Android 6.0 (API 23).
+
+In the Android developer documentation, a reference implementation can be found on how to [Show a biometric authentication dialog](https://developer.android.com/training/sign-in/biometric-auth "Show a biometric authentication dialog").
+
+There are two `authenticate` methods available in the `BiometricPrompt` class. The difference is that one method is expecting a [CryptoObject](https://developer.android.com/reference/android/hardware/biometrics/BiometricPrompt.CryptoObject.html "CryptoObject"), which adds an additional layer of security for the biometric authentication.
+
+The authentication flow would be as follows when using CryptoObject:
+
+- The app creates a key in the KeyStore with `setUserAuthenticationRequired` set to true and `setInvalidatedByBiometricEnrollment` set to -1.
+- This key is used to encrypt information that is authenticating the user (e.g. session information or authentication token).
+- A valid set of biometrics must be presented before the key is released from the KeyStore to decrypt the data, which is validated through the authenticate method and the CryptoObject.
+- This solution cannot be bypassed, even on rooted devices, as the key from the KeyStore can only be used after successful biometric authentication.
+
+If CryptoObject is not used as part of the authenticate method, it can be bypassed by using Frida. See the Dynamic Instrumentation section for more details.

```

Comment:
```suggestion
If `CryptoObject` is not used as part of the authenticate method, it can be bypassed by using Frida. See the "Dynamic Instrumentation" section for more details.
```